File: backend/internal/service/subsite.go
package service

import "time"

const (
	SubSiteStatusPending  = "pending"
	SubSiteStatusActive   = "active"
	SubSiteStatusDisabled = "disabled"
)

const (
	SubSiteRegistrationOpen   = "open"
	SubSiteRegistrationInvite = "invite"
	SubSiteRegistrationClosed = "closed"
)

const (
	SubSiteThemeTemplateStarter  = "starter"
	SubSiteThemeTemplateAurora   = "aurora"
	SubSiteThemeTemplateSummit   = "summit"
	SubSiteThemeTemplateTerminal = "terminal"
)

const (
	DefaultSubSiteActivationPriceFen = 38800
	DefaultSubSiteValidityDays       = 365
	DefaultSubSiteMaxLevel           = 2
	MaxSubSiteLevelHardLimit         = 10
	DefaultSubSiteConsumeRate        = 1.0
	MaxSubSiteConsumeRateMultiplier  = 10.0
)

// 分站模式：
//   pool — 经典资金池模式：分站主先向主站充值到 balance_fen，线下卖用户余额赚差价；
//          支持配置自有收款账号，用户线上充值直接到分站主账号并从池同步扣减（自动进货）。
//   rate — 倍率分成模式：无独立资金池；用户充值/消费都经主站，消费时按链上复合倍率
//          把 (compound_i - compound_{i-1}) × base 的利润分级入账到 balance_fen，
//          分站主通过提现把利润取走。
const (
	SubSiteModePool = "pool"
	SubSiteModeRate = "rate"
)

type SubSiteThemeTemplateOption struct {
	Key         string `json:"key"`
	Label       string `json:"label"`
	Description string `json:"description"`
}

var DefaultSubSiteThemeTemplates = []SubSiteThemeTemplateOption{
	{
		Key:         SubSiteThemeTemplateStarter,
		Label:       "Starter",
		Description: "通用 SaaS 风格，适合大多数分站直接开箱使用",
	},
	{
		Key:         SubSiteThemeTemplateAurora,
		Label:       "Aurora",
		Description: "更强展示感的渐变品牌模板，适合营销导向站点",
	},
	{
		Key:         SubSiteThemeTemplateSummit,
		Label:       "Summit",
		Description: "偏企业感的信息型模板，适合强调稳定与专业服务",
	},
	{
		Key:         SubSiteThemeTemplateTerminal,
		Label:       "Terminal",
		Description: "偏开发者风格模板，适合 Claude Code / Codex 场景",
	},
}

type SubSiteOpenInfo struct {
	Enabled              bool                         `json:"enabled"`
	Scope                string                       `json:"scope"`
	ParentSubSiteID      *int64                       `json:"parent_sub_site_id,omitempty"`
	ParentSubSiteName    string                       `json:"parent_sub_site_name,omitempty"`
	Level                int                          `json:"level"`
	MaxLevel             int                          `json:"max_level"`
	PriceFen             int                          `json:"price_fen"`
	ValidityDays         int                          `json:"validity_days"`
	Currency             string                       `json:"currency"`
	AllowCustomDomain    bool                         `json:"allow_custom_domain"`
	DefaultThemeTemplate string                       `json:"default_theme_template"`
	ThemeTemplates       []SubSiteThemeTemplateOption `json:"theme_templates"`
}

type PlatformSubSiteConfig struct {
	EntryEnabled         bool                         `json:"entry_enabled"`
	Enabled              bool                         `json:"enabled"`
	ActivationPriceFen   int                          `json:"activation_price_fen"`
	ValidityDays         int                          `json:"validity_days"`
	MaxLevel             int                          `json:"max_level"`
	DefaultThemeTemplate string                       `json:"default_theme_template"`
	ThemeTemplates       []SubSiteThemeTemplateOption `json:"theme_templates"`
}

type UpdatePlatformSubSiteConfigInput struct {
	EntryEnabled         bool   `json:"entry_enabled"`
	Enabled              bool   `json:"enabled"`
	ActivationPriceFen   int    `json:"activation_price_fen"`
	ValidityDays         int    `json:"validity_days"`
	MaxLevel             int    `json:"max_level"`
	DefaultThemeTemplate string `json:"default_theme_template"`
}

type SubSite struct {
	ID                    int64               `json:"id"`
	OwnerUserID           int64               `json:"owner_user_id"`
	OwnerEmail            string              `json:"owner_email,omitempty"`
	ParentSubSiteID       *int64              `json:"parent_sub_site_id,omitempty"`
	ParentSubSiteName     string              `json:"parent_sub_site_name,omitempty"`
	Level                 int                 `json:"level"`
	Name                  string              `json:"name"`
	Slug                  string              `json:"slug"`
	CustomDomain          string              `json:"custom_domain,omitempty"`
	Status                string              `json:"status"`
	Mode                  string              `json:"mode"`
	SiteLogo              string              `json:"site_logo,omitempty"`
	SiteFavicon           string              `json:"site_favicon,omitempty"`
	SiteSubtitle          string              `json:"site_subtitle,omitempty"`
	Announcement          string              `json:"announcement,omitempty"`
	ContactInfo           string              `json:"contact_info,omitempty"`
	DocURL                string              `json:"doc_url,omitempty"`
	HomeContent           string              `json:"home_content,omitempty"`
	ThemeTemplate         string              `json:"theme_template,omitempty"`
	RegistrationMode      string              `json:"registration_mode,omitempty"`
	EnableTopup           bool                `json:"enable_topup"`
	AllowSubSite          bool                `json:"allow_sub_site"`
	SubSitePriceFen       int                 `json:"sub_site_price_fen"`
	ConsumeRateMultiplier float64             `json:"consume_rate_multiplier"`
	BalanceFen            int64               `json:"balance_fen"`
	TotalTopupFen         int64               `json:"total_topup_fen"`
	TotalConsumedFen      int64               `json:"total_consumed_fen"`
	TotalWithdrawnFen     int64               `json:"total_withdrawn_fen"`
	AllowOnlineTopup      bool                `json:"allow_online_topup"`
	AllowOfflineTopup     bool                `json:"allow_offline_topup"`
	OwnerPaymentConfig    *OwnerPaymentConfig `json:"owner_payment_config,omitempty"`
	SubscriptionExpiredAt *time.Time          `json:"subscription_expired_at,omitempty"`
	UserCount             int64               `json:"user_count,omitempty"`
	ChildSiteCount        int64               `json:"child_site_count,omitempty"`
	EntryURL              string              `json:"entry_url,omitempty"`
	CreatedAt             time.Time           `json:"created_at"`
	UpdatedAt             time.Time           `json:"updated_at"`
}

// OwnerPaymentConfig 分站主自有收款凭据（仅 pool 模式可用）。
// 任一渠道启用时对应 Enabled 字段为 true；未配置（nil 或 Enabled=false）则退回全局 settings。
type OwnerPaymentConfig struct {
	Wechat *WechatPayCredentials `json:"wechat,omitempty"`
	Alipay *AlipayCredentials    `json:"alipay,omitempty"`
	Epay   *EpayCredentials      `json:"epay,omitempty"`
}

type WechatPayCredentials struct {
	Enabled      bool   `json:"enabled"`
	AppID        string `json:"app_id"`
	MchID        string `json:"mch_id"`
	APIv3Key     string `json:"apiv3_key"`
	MchSerialNo  string `json:"mch_serial_no"`
	PublicKeyID  string `json:"public_key_id"`
	PublicKey    string `json:"public_key"`
	PrivateKey   string `json:"private_key"`
	NotifyURL    string `json:"notify_url"`
}

type AlipayCredentials struct {
	Enabled      bool   `json:"enabled"`
	AppID        string `json:"app_id"`
	PrivateKey   string `json:"private_key"`
	PublicKey    string `json:"public_key"`
	NotifyURL    string `json:"notify_url"`
	IsProduction bool   `json:"is_production"`
}

type EpayCredentials struct {
	Enabled   bool   `json:"enabled"`
	Gateway   string `json:"gateway"`
	PID       string `json:"pid"`
	PKey      string `json:"pkey"`
	NotifyURL string `json:"notify_url"`
}

type CreateSubSiteInput struct {
	OwnerUserID           int64               `json:"owner_user_id"`
	ParentSubSiteID       *int64              `json:"parent_sub_site_id,omitempty"`
	Name                  string              `json:"name"`
	Slug                  string              `json:"slug"`
	CustomDomain          string              `json:"custom_domain"`
	Status                string              `json:"status"`
	Mode                  string              `json:"mode"`
	SiteLogo              string              `json:"site_logo"`
	SiteFavicon           string              `json:"site_favicon"`
	SiteSubtitle          string              `json:"site_subtitle"`
	Announcement          string              `json:"announcement"`
	ContactInfo           string              `json:"contact_info"`
	DocURL                string              `json:"doc_url"`
	HomeContent           string              `json:"home_content"`
	ThemeTemplate         string              `json:"theme_template"`
	RegistrationMode      string              `json:"registration_mode"`
	EnableTopup           *bool               `json:"enable_topup,omitempty"`
	AllowSubSite          *bool               `json:"allow_sub_site,omitempty"`
	SubSitePriceFen       int                 `json:"sub_site_price_fen"`
	ConsumeRateMultiplier float64             `json:"consume_rate_multiplier"`
	AllowOnlineTopup      *bool               `json:"allow_online_topup,omitempty"`
	AllowOfflineTopup     *bool               `json:"allow_offline_topup,omitempty"`
	OwnerPaymentConfig    *OwnerPaymentConfig `json:"owner_payment_config,omitempty"`
	SubscriptionExpiredAt *time.Time          `json:"subscription_expired_at,omitempty"`
}

// UpdateSubSiteInput 平台管理员可修改的完整字段集（含 Mode / SubscriptionExpiredAt 等敏感字段）。
// 分站主自助路径请使用 UpdateOwnedSubSiteInput，字段范围更窄。
type UpdateSubSiteInput struct {
	ID                    int64               `json:"id"`
	OwnerUserID           int64               `json:"owner_user_id"`
	ParentSubSiteID       *int64              `json:"parent_sub_site_id,omitempty"`
	Name                  string              `json:"name"`
	Slug                  string              `json:"slug"`
	CustomDomain          string              `json:"custom_domain"`
	Status                string              `json:"status"`
	Mode                  string              `json:"mode"`
	SiteLogo              string              `json:"site_logo"`
	SiteFavicon           string              `json:"site_favicon"`
	SiteSubtitle          string              `json:"site_subtitle"`
	Announcement          string              `json:"announcement"`
	ContactInfo           string              `json:"contact_info"`
	DocURL                string              `json:"doc_url"`
	HomeContent           string              `json:"home_content"`
	ThemeTemplate         string              `json:"theme_template"`
	RegistrationMode      string              `json:"registration_mode"`
	EnableTopup           *bool               `json:"enable_topup,omitempty"`
	AllowSubSite          *bool               `json:"allow_sub_site,omitempty"`
	SubSitePriceFen       int                 `json:"sub_site_price_fen"`
	ConsumeRateMultiplier float64             `json:"consume_rate_multiplier"`
	AllowOnlineTopup      *bool               `json:"allow_online_topup,omitempty"`
	AllowOfflineTopup     *bool               `json:"allow_offline_topup,omitempty"`
	OwnerPaymentConfig    *OwnerPaymentConfig `json:"owner_payment_config,omitempty"`
	SubscriptionExpiredAt *time.Time          `json:"subscription_expired_at,omitempty"`
}

// UpdateOwnedSubSiteInput 分站主自助可改字段；敏感字段（Mode/Status/ParentSubSiteID/
// SubscriptionExpiredAt/AllowSubSite/AllowOnlineTopup/EnableTopup/Level）由 admin 接口管理。
type UpdateOwnedSubSiteInput struct {
	ID                    int64               `json:"id"`
	Name                  string              `json:"name"`
	Slug                  string              `json:"slug"`
	CustomDomain          string              `json:"custom_domain"`
	SiteLogo              string              `json:"site_logo"`
	SiteFavicon           string              `json:"site_favicon"`
	SiteSubtitle          string              `json:"site_subtitle"`
	Announcement          string              `json:"announcement"`
	ContactInfo           string              `json:"contact_info"`
	DocURL                string              `json:"doc_url"`
	HomeContent           string              `json:"home_content"`
	ThemeTemplate         string              `json:"theme_template"`
	RegistrationMode      string              `json:"registration_mode"`
	SubSitePriceFen       int                 `json:"sub_site_price_fen"`
	ConsumeRateMultiplier float64             `json:"consume_rate_multiplier"`
	AllowOfflineTopup     *bool               `json:"allow_offline_topup,omitempty"`
	OwnerPaymentConfig    *OwnerPaymentConfig `json:"owner_payment_config,omitempty"`
}

type CreateSubSiteActivationInput struct {
	Name                  string  `json:"name"`
	Slug                  string  `json:"slug"`
	CustomDomain          string  `json:"custom_domain"`
	SiteLogo              string  `json:"site_logo"`
	SiteFavicon           string  `json:"site_favicon"`
	SiteSubtitle          string  `json:"site_subtitle"`
	Announcement          string  `json:"announcement"`
	ContactInfo           string  `json:"contact_info"`
	DocURL                string  `json:"doc_url"`
	HomeContent           string  `json:"home_content"`
	ThemeTemplate         string  `json:"theme_template"`
	RegistrationMode      string  `json:"registration_mode"`
	EnableTopup           *bool   `json:"enable_topup,omitempty"`
	AllowSubSite          *bool   `json:"allow_sub_site,omitempty"`
	SubSitePriceFen       int     `json:"sub_site_price_fen"`
	ConsumeRateMultiplier float64 `json:"consume_rate_multiplier"`
}

type SubSiteActivationRequest struct {
	PaymentOrderID     int64                        `json:"payment_order_id"`
	UserID             int64                        `json:"user_id"`
	ParentSubSiteID    *int64                       `json:"parent_sub_site_id,omitempty"`
	Level              int                          `json:"level"`
	ValidityDays       int                          `json:"validity_days"`
	Site               CreateSubSiteActivationInput `json:"site"`
	ActivatedSubSiteID *int64                       `json:"activated_sub_site_id,omitempty"`
	ActivatedAt        *time.Time                   `json:"activated_at,omitempty"`
	CreatedAt          time.Time                    `json:"created_at"`
	UpdatedAt          time.Time                    `json:"updated_at"`
}

File: frontend/src/api/subsite.ts
import { apiClient } from './client'
import type { CreateSubSiteActivationInput } from './payment'
import type { PaginatedResponse } from '@/types'

export interface SubSiteThemeTemplateOption {
  key: string
  label: string
  description: string
}

export interface SubSiteOpenInfo {
  enabled: boolean
  scope: 'platform' | 'subsite'
  parent_sub_site_id?: number
  parent_sub_site_name?: string
  level: number
  max_level: number
  price_fen: number
  validity_days: number
  currency: string
  allow_custom_domain: boolean
  default_theme_template: string
  theme_templates: SubSiteThemeTemplateOption[]
}

export interface OwnedSubSite {
  id: number
  owner_user_id: number
  owner_email?: string
  parent_sub_site_id?: number
  parent_sub_site_name?: string
  level: number
  name: string
  slug: string
  custom_domain?: string
  status: 'pending' | 'active' | 'disabled'
  mode: 'pool' | 'rate'
  site_logo?: string
  site_favicon?: string
  site_subtitle?: string
  announcement?: string
  contact_info?: string
  doc_url?: string
  home_content?: string
  pending_home_content?: string
  home_content_review_status?: 'none' | 'pending' | 'approved' | 'rejected'
  home_content_review_note?: string
  home_content_submitted_at?: string
  home_content_reviewed_at?: string
  home_content_reviewed_by?: number
  theme_template?: string
  registration_mode?: 'open' | 'invite' | 'closed'
  enable_topup: boolean
  allow_sub_site: boolean
  sub_site_price_fen: number
  consume_rate_multiplier?: number
  subscription_expired_at?: string | null
  user_count?: number
  child_site_count?: number
  entry_url?: string
  balance_fen?: number
  total_topup_fen?: number
  total_consumed_fen?: number
  total_withdrawn_fen?: number
  allow_online_topup?: boolean
  allow_offline_topup?: boolean
  owner_payment_config?: OwnerPaymentConfig | null
  created_at: string
  updated_at: string
}

export interface OwnerPaymentConfig {
  wechat?: { enabled: boolean; app_id: string; mch_id: string; apiv3_key: string; mch_serial_no: string; public_key_id: string; public_key: string; private_key: string; notify_url: string }
  alipay?: { enabled: boolean; app_id: string; private_key: string; public_key: string; notify_url: string; is_production: boolean }
  epay?: { enabled: boolean; gateway: string; pid: string; pkey: string; notify_url: string }
}

export interface OwnedSubSiteLedgerEntry {
  id: number
  sub_site_id: number
  tx_type: string
  delta_fen: number
  balance_after_fen: number
  related_user_id?: number
  related_usage_log_id?: number
  related_order_id?: number
  operator_id?: number
  note?: string
  created_at: string
}

export type UpdateOwnedSubSiteRequest = CreateSubSiteActivationInput

export async function getOpenInfo(): Promise<SubSiteOpenInfo> {
  const { data } = await apiClient.get<SubSiteOpenInfo>('/subsite/open-info')
  return {
    ...data,
    theme_templates: Array.isArray(data?.theme_templates) ? data.theme_templates : [],
  }
}

export async function listOwnedSites(): Promise<OwnedSubSite[]> {
  const { data } = await apiClient.get<OwnedSubSite[] | null>('/subsite/owned')
  if (!Array.isArray(data)) {
    return []
  }
  return data
}

export async function getOwnedSite(id: number): Promise<OwnedSubSite> {
  const { data } = await apiClient.get<OwnedSubSite>(`/subsite/owned/${id}`)
  return data
}

export async function updateOwnedSite(id: number, payload: UpdateOwnedSubSiteRequest): Promise<OwnedSubSite> {
  const { data } = await apiClient.put<OwnedSubSite>(`/subsite/owned/${id}`, payload)
  return data
}

export async function offlineTopupUser(
  siteID: number,
  payload: { user_id: number; amount_fen: number; note?: string }
): Promise<OwnedSubSite> {
  const { data } = await apiClient.post<OwnedSubSite>(`/subsite/owned/${siteID}/offline-topup`, payload)
  return data
}

export async function listOwnedLedger(
  siteID: number,
  page: number = 1,
  pageSize: number = 20,
  txType?: string
): Promise<PaginatedResponse<OwnedSubSiteLedgerEntry>> {
  const { data } = await apiClient.get<PaginatedResponse<OwnedSubSiteLedgerEntry>>(
    `/subsite/owned/${siteID}/ledger`,
    {
      params: {
        page,
        page_size: pageSize,
        tx_type: txType || undefined
      }
    }
  )
  return data
}

export async function createPoolTopupOrder(payload: {
  site_id: number
  amount_fen: number
  pay_method?: string
}): Promise<{
  order_no: string
  code_url?: string
  amount_fen: number
  expired_at: string
}> {
  const { data } = await apiClient.post('/payment/subsite-topup', payload)
  return data as any
}

export const subSiteAPI = {
  getOpenInfo,
  listOwnedSites,
  getOwnedSite,
  updateOwnedSite,
  offlineTopupUser,
  listOwnedLedger,
  createPoolTopupOrder,
}

export default subSiteAPI
